Add `noselect` to `wildmode` option

girishji opened the pull request:
When `noselect` is present in `wildmode` and `wildmenu` is enabled, the completion menu appears without pre-selecting the first item.

This change makes it easier to implement command-line **auto-completion**, where the menu dynamically appears as characters are typed, and `<Tab>` can be used to manually select an item. This can be achieved by leveraging the `CmdlineChanged` event to insert `wildchar(m)`, triggering completion menu.

Without this change, auto-completion using the `wildmenu` mechanism is not feasible, as it automatically inserts the first match, preventing dynamic selection.

The following snippet demonstrates how to configure auto-completion using `noselect`:

```vim
set wim=noselect:lastused,full wop=pum wcm=<C-@> wmnu
autocmd CmdlineChanged : timer_start(0, function(CmdComplete, [getcmdline()]))
def CmdComplete(cur_cmdline: string, timer: number)
    var [cmdline, curpos] = [getcmdline(), getcmdpos()]
    if cur_cmdline ==# cmdline # Avoid completing each char of keymaps and pasted text
      && !pumvisible() && curpos == cmdline->len() + 1
      && cmdline =~ '\%(\w\|[*/:.-]\)$' && cmdline !~ '^\d\+$'  # Reduce noise
        feedkeys("\<C-@>", "ti")
        set eventignore+=CmdlineChanged  # Suppress redundant completion attempts
        timer_start(0, (_) => {
            getcmdline()->substitute('\%x00', '', 'g')->setcmdline()  # Remove <C-@> if no completion items exist
            set eventignore-=CmdlineChanged
        })
    endif
enddef
autocmd CmdlineEnter : set bo+=error
autocmd CmdlineLeave : set bo-=error

def MuteEventAndSend(key: string): string
    set ei+=CmdlineChanged
    timer_start(0, (_) => execute('set ei-=CmdlineChanged'))
    return (pumvisible() ? "\<c-e>" : "") .. key
enddef
cnoremap <expr> <up> MuteEventAndSend("\<up>")
cnoremap <expr> <down> MuteEventAndSend("\<down>")

```

---

**Edit:**

Here are a few examples that utilize the above auto-completion mechanism:

**Fuzzy Find:**

```vim
set findfunc=FuzzyFind
def FuzzyFind(cmdarg: string, _: bool): list<string>
    if allfiles == null_list
        allfiles = systemlist($'find {get(g:, "fzfind_root", ".")} \! \( -path "*/.git" -prune \) -type f -follow')
    endif
    return cmdarg == '' ? allfiles : allfiles->matchfuzzy(cmdarg)
enddef
var allfiles: list<string>
autocmd CmdlineEnter : allfiles = null_list
nnoremap <leader>f :find<space><c-@>
nnoremap <leader>V :<c-r>=execute('let fzfind_root="$VIMRUNTIME"')\|''<cr>find<space><c-@>
```

**Grep (Interactive):**

```vim
command! -nargs=+ -complete=customlist,GrepComplete Grep VisitFile(<q-args>)
def GrepComplete(arglead: string, cmdline: string, cursorpos: number): list<any>
    return arglead != null_string ? systemlist($'grep -REIHns "{arglead}"' ..
       ' --exclude-dir=.git --exclude=".*"') : []
enddef
def VisitFile(line: string)
    var qfitem = getqflist({lines: [line]}).items[0]
    if qfitem->has_key('bufnr')
        var pos = qfitem.vcol > 0 ? 'setcharpos' : 'setpos'
        exec $':b +call\ {pos}(".",\ [0,\ {qfitem.lnum},\ {qfitem.col},\ 0]) {qfitem.bufnr}'
        setbufvar(qfitem.bufnr, '&buflisted', 1)
    endif
enddef
nnoremap <leader>g :Grep<space>
nnoremap <leader>G :Grep <c-r>=expand("<cword>")<cr>

```

girishji committed:
Add `noselect` to `wildmode` option

When `noselect` is present in `wildmode` and 'wildmenu' is enabled, the
completion menu appears without pre-selecting the first item.

This change makes it easier to implement command-line auto-completion, where
the menu dynamically appears as characters are typed, and `<Tab>` can be used
to manually select an item. This can be achieved by leveraging the `CmdlineChanged`
event to insert `wildchar(m)`, triggering completion menu.

Without this change, auto-completion using the 'wildmenu' mechanism is not
feasible, as it automatically inserts the first match, preventing dynamic
selection.

The following Vimscript snippet demonstrates how to configure auto-completion
using `noselect`:

```vim
set wim=noselect:lastused,full wop=pum wcm=<C-@> wmnu
autocmd CmdlineChanged : timer_start(0, function(CmdComplete, [getcmdline()]))

def CmdComplete(cur_cmdline: string, timer: number)
    var [cmdline, curpos] = [getcmdline(), getcmdpos()]

    if cur_cmdline ==# cmdline  # Avoid completing each character in keymaps and pasted text
        && !pumvisible() && curpos == cmdline->len() + 1

        if cmdline[curpos - 2] =~ '[\w*/:]'  # Reduce noise by completing only selected characters
            feedkeys("\<C-@>", "ti")
            set eventignore+=CmdlineChanged  # Suppress redundant completion attempts

            timer_start(0, (_) => {
                getcmdline()->substitute('\%x00$', '', '')->setcmdline()  # Remove <C-@> if no completion items exist
                set eventignore-=CmdlineChanged
            })
        endif
    endif
enddef
```

M  runtime/doc/options.txt
M  runtime/doc/tags
M  src/cmdexpand.c
M  src/ex_getln.c
M  src/option.h
M  src/testdir/gen_opt_test.vim
M  src/testdir/test_cmdline.vim
M  src/vim.h

girishji committed:
Put back a line removed in last commit

M  runtime/doc/tags

girishji committed:
Add tests to cover menu with single item

M  src/testdir/test_cmdline.vim

zeertzjq commented on src/testdir/test_cmdline.vim:
```diff
@@ -2042,6 +2042,34 @@ func Wildmode_tests()
+  set wildmode=noselect:longest
+  call feedkeys(":MyCmd o\t\<C-B>\"\<CR>", 'xt')
+  call assert_equal('"MyCmd one', @:)
+  set wildmenu&
```
Tthe default value of `'wildmenu'` is on:
```c
    {"wildmenu",    "wmnu", P_BOOL|P_VI_DEF,
			    (char_u *)&p_wmnu, PV_NONE, NULL, NULL,
			    {(char_u *)TRUE, (char_u *)0L} SCTX_INIT},
```
And `'wildmenu'` is already on here, so this line doesn't do anything.

girishji replied:
Fixed.

zeertzjq commented on src/testdir/test_cmdline.vim:
```diff
@@ -2071,6 +2099,21 @@ func Wildmode_tests()
+  call assert_equal('"MyCmd o', @:)
+  call feedkeys(":MyCmd o\t\t\<C-B>\"\<CR>", 'xt')
+  call assert_equal('"MyCmd oneA', @:)
+  set wildmenu&
```
Same here. This line doesn't do anything.

girishji committed:
Remove redundant settings

AD src/testdir/XtestDebug.vim
M  src/testdir/test_cmdline.vim
A  src/testdir/vimcmd

girishji committed:
Fix test

D  src/testdir/XtestDebug.vim
M  src/testdir/test_cmdline.vim
D  src/testdir/vimcmd

girishji committed:
Improve doc

M  src/testdir/test_cmdline.vim

ychin commented on src/ex_getln.c:
```diff
@@ -2716,6 +2720,8 @@ check_opt_wim(void)
 	    new_wim_flags[idx] |= WIM_LIST;
 	else if (i == 8 && STRNCMP(p, "lastused", 8) == 0)
 	    new_wim_flags[idx] |= WIM_BUFLASTUSED;
```
Per the code comment above, please keep this in sync with `p_wim_values` and add this to that array. You won't get auto-complete otherwise.

girishji committed:
Make auto-complete menu include 'noselect'

M  src/optionstr.c

girishji committed:
Add test for completion of 'noselect' value

M  src/testdir/test_cmdline.vim

girishji committed:
Fix <C-Y> crash
Caused by segfault, when item gets deleted
from completion list before being accessed.

M  src/ex_getln.c
M  src/testdir/test_cmdline.vim

zeertzjq commented on src/ex_getln.c:
```diff
@@ -974,7 +977,7 @@ cmdline_wildchar_complete(
 
 		    // remove match
-		    nextwild(xp, WILD_PREV, 0, escape);
+		    nextwild(xp, WILD_PREV, 0 | (options & ~WIM_NOSELECT), escape);
```
This doesn't look right. `WIM_NOSELECT` is `0x10`, which is equal to `WILD_ADD_SLASH`, which is never set in `options`.

zeertzjq replied:
And the `0 | ` isn't needed here. But why do `options` now need to be passed when they don't need to be passed before?

ychin replied:
I was wondering about that too

girishji replied:
Good catch. It was supposed to be just `nextwild(xp, WILD_PREV, options, escape);` (replace `0` with `options`). Don't know how the meaningless thing got in there. `options` is needed here to prevent xp_files from being emptied out if there is only 1 entry. For `noselect` single element completion menu should be retained. `options` contains `WILD_KEEP_SOLE_ITEM` when `noselect` is set.

girishji replied:
Fixed https://github.com/vim/vim/pull/16778

justinmk commented on src/testdir/test_cmdline.vim:
```diff
@@ -2042,16 +2042,52 @@ func Wildmode_tests()
+  call assert_equal('"MyCmd o', @:)
+
+  " When 'full' is present, complete after first <tab>.
+  set wildmode=noselect,full
```
Seems to be order-sensitive. Does the test pass if this line is changed to:

    set wildmode=full,noselect

zeertzjq replied:
Isn't this just how `'wildmode'` works?

justinmk replied:
That means `set wildmode+=noselect` doesn't work with the default value. Seems strange, and probably should be documented if this is expected.

ychin replied:
Doesn't the docs say this? 

> It is a comma-separated list of up to four parts.  Each part specifies what to do for each consecutive use of 'wildchar'.  The first part specifies the behavior for the first use of 'wildchar', The second part for the second use, etc.

> Each part consists of a colon separated list consisting of the following possible values:
> …

It has always been order sensitive and the examples also don't use `+=`. I do wonder if it `noselect` ever makes sense to be not used as the first item in the element.

justinmk replied:
No, not really. The docs you quoted are maybe barely understandable in these terms *if* you already know what you are looking for.

If `set wildmode+=...` produces nonsense it should be an error, instead of pretending to support `+=`.

ychin replied:
Sure, I'm just pointing out that it's not specific to this `noselect` change, as the other values there are already order dependent for this option. That part of the docs has always been a bit hard to parse. I tried to give more examples in #8078 but maybe it's not clear enough.

I feel like clearing up the docs is better than just making `+=` produce an error though, since it's not like it literally produces nonsense (unless you mean doing extra validation inside `set wildmode+=` to check each value). It just adds a value to the end. The important part is understanding that `wildmode` is order dependent (it's kind of impossible to properly configure this option without understanding that).

girishji replied:
The `wildmode` option is inherently order-sensitive. Each comma-separated fragment determines the sequence in which `wildchar` behaves. Within each fragment, the colon-separated parts define simultaneous effects for `wildchar`. This is not clearly explained in the documentation. I considered rewriting the docs but chose not to include it in this PR.  

Additionally, `set wildmode=full,noselect` does not behave as expected. Ideally, the first `wildchar` should open the menu, selecting and inserting the first item (`full`), while the second `wildchar` should unselect the item and revert the command-line change (`noselect`). However, this use case is unlikely to be meaningful. In practice, the implementation simply ignores `noselect` unless it appears before the first comma.

girishji replied:
https://github.com/vim/vim/pull/17100

SALEH55155 reviewed (approved):
76cc4a0d8c96d87514da1d00cbec42172b21301b
